CONTRIBUTING.md

```diff
@@ -0,0 +1,458 @@
+    via the "Apply Suggested Changes" GitHub workflow, and then to amend the
+    resulting commit to fix any related compilation or test errors or
+    formatting/lint-related changes; this ensures that correct co-author
+    metadata is included in the commit. If the changes are substantial enough
+    that it makes more sense to rewrite the original commit, make sure to
+    include co-author metadata in the commit message when doing so (squashing
+    the GitHub-generate suggestion acceptance commit(s) together with the
+    original commit in an interactive rebase can make this easy).
+
+### Coding Style
+
+The `librustzcash` authors hold our software to a high standard of quality. The
+list of style requirements below is not comprehensive, but violation of any of
+the following guidelines is likely to cause your pull request to be rejected or
+changes to be required. The coding style in this repository has evolved over
+time, and not all preexisting code follows this style; when modifications are
+being made to existing code, it should be upgraded to reflect the recommended
+style (although please ensure that you separate functional changes from
+style-oriented refactoring in the Git commit history.)
+
+#### Type Safety
+
+In `librustzcash` code, type safety is of paramount importance. This has
+numerous implications, including but not limited to the following:
+- Invalid states should be made unrepresentable at the type level. In general:
+  - `structs` should have all internal members private or crate-private, and
+    should expose constructors that result in `Result<...>` or `Option<...>`
+    that check for invariant violations, if any such violations are possible.
+    Provide public or crate-public accessors for internal members when necessary.
+  - "bare" native integer types, strings, and so forth should be avoided in
+    public APIs; use "newtype" wrappers with clearly documented semantics instead.
+  - Avoid platform-specific integer sizing (i.e. `usize`) except when e.g.
+    indexing into a Rust collection type that already requires such semantics.
+  - Use `enum`s liberally; a common type safety failure in many other languages
+    is that product (struct or tuple) types containing potentially invalid
+    state space are used.
+  - Use custom `enum`s with semantically relevant variants instead of boolean
+    arguments and return values.
+- Prefer immutability; make data types immutable unless there is a strong
+  reason to believe that values will need to be modified in-place for
+  performance reasons.
+- Take care when introducing and/or using structured enum variants, because
+  Rust does not provide adequate language features for making such values
+  immutable or ensuring safe construction. Instead of creating structured or
+  tuple variants, it is often preferable for a variant to wrap an immutable
+  type and expose a safe constructor for the variant along with accessors for
+  the members of the wrapped type.
```

Comment:
I am guessing that an 'immutable type' means a type with a private field that doesn't expose a method to mutate it freely, but I think this is hard to parse. Something like?
```suggestion
- Take care when introducing and/or using structured or tuple enum variants, because
  Rust does not provide adequate language features for making such values
  immutable or ensuring safe construction. Instead of creating structured or
  tuple variants that directly contain primitive types, it is often preferable to define
  and wrap a new type with private fields, and expose a safe constructor for the 
  variant along with accessors for the members of the wrapped type.
```